samza-api/src/test/java/org/apache/samza/operators/descriptors/TestTransformingInputDescriptor.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.descriptors;
+
+import java.util.Collections;
+import org.apache.samza.operators.descriptors.transforming.MockTransformingInputDescriptor;
+import org.apache.samza.operators.descriptors.transforming.MockTransformingOutputDescriptor;
+import org.apache.samza.operators.descriptors.transforming.MockTransformingSystemDescriptor;
+import org.apache.samza.serializers.IntegerSerde;
+import org.apache.samza.system.SystemStreamMetadata;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class TestTransformingInputDescriptor {
```

Comment:
System descriptors can provide both a transform and expand function at the same time. For an example, see MockExpandingSystemDescriptor. The transform function is called when messages are delivered to the InputOperatorImpl for the input descriptor. Expand function is called during describe/planning stage when calling getInputStream.

In the current implementation, an expanding system descriptor will still implement the TransformingInputDescriptorProvider interface. This is because for the current case of a single getInputStream(String, Serde) variant, the method signature for transforming and expanding descriptors is the same, hence the current interface definition is also the same. I'd be OK with adding the duplicate interface; it does seem semantically clearer, and will be useful if we decide to add support for stream level transform functions. What do you think?

Reply:
As discussed offline, agreed with the current approach.